app/models/saved_page.rb

```diff
@@ -0,0 +1,26 @@
+class SavedPage < ApplicationRecord
+  belongs_to :oidc_user
+
+  validates :page_path, uniqueness: { scope: :oidc_user_id }, presence: true
+
+  validate :page_path_is_valid_path
+
+  def to_hash
+    {
+      "page_path" => page_path,
+    }
+  end
+
+private
+
+  def page_path_is_valid_path
+    errors.add(:page_path, :invalid_path, message: "must only include URL path") if is_invalid_path?
+  end
+
+  def is_invalid_path?
+    return true if page_path.blank?
+
+    uri = URI(page_path)
```

Comment:
Can we make this a bit more robust to bad input?  Currently it'll throw an exception if given something that doesn't look like a URI:

```
irb(main):010:0> URI("hello world")
Traceback (most recent call last):
        1: from (irb):10
URI::InvalidURIError (bad URI(is not URI?): "hello world")

irb(main):011:0> URI(nil)
Traceback (most recent call last):
        2: from (irb):10
        1: from (irb):11:in `rescue in irb_binding'
ArgumentError (bad argument (expected URI object or URI string))
```

Reply:
That's true, however I think blank / empty is the main case here and that guard clause on L21 covers that. Plus they'd get a warning about `can't be blank` from the rails validations too.

I've tested a bunch of these and URI seems pretty open:

```
URI("https://gov.uk/foo")                                                                                                                                                                         
=> #<URI::HTTPS https://gov.uk/foo>

URI("www.gov.uk/foo")                                                                                                                                                                             
=> #<URI::Generic www.gov.uk/foo>

URI("gov.uk/foo")                                                                                                                                                                                 
=> #<URI::Generic gov.uk/foo>

URI("/foo")                                                                                                                                                                                       
=> #<URI::Generic /foo>

URI("foo")                                                                                                                                                                                        
=> #<URI::Generic foo>
```

If it is none of those and not nil, then I kind of think throwing an error might be good!

Reply:
It actually also works with 
```
URI("")                                                                                                                                                                                           
=> #<URI::Generic >
```

... which tbh seems pretty bad

Reply:
> If it is none of those and not nil, then I kind of think throwing an error might be good!

It should be a validation error we return to the user, not a 5xx error due to our validation code throwing an uncaught exception

Reply:
I looked into other ways of doing it, URI comes with some regex that you can use for comparisons.
However here we don't really care about the rest of the URI. it's just if it can be parsed into URI.path and if the output matches

Reply:
How about

```ruby
def is_valid_path?(page_path)
  URI(page_path) == page_path
rescue StandardError
  false
end
```

Then change line 17 to use `unless is_valid_path?`

Reply:
oh actually turns out `URI("foo").path` becomes `foo` so I might just add .starts_with("/") as a condition too